File: src/main/java/com/learning/leetcode/arrays/easy/BestTimeBuySellStock.py
import sys


def maxProfit(prices):
    min_price = sys.maxsize
    profit = 0
    for price in prices:
        min_price = min(min_price, price)
        profit = max(profit, price - min_price)
    return profit


def main():
    prices = [7, 1, 5, 3, 6, 4]
    profit = maxProfit(prices)
    print("Max profit for " + str(prices) + " is " + str(profit))
    prices = [7, 6, 4, 3, 1]
    profit = maxProfit(prices)
    print("Max profit for " + str(prices) + " is " + str(profit))


main()

File: src/main/java/com/learning/leetcode/arrays/easy/BestTimeBuySellStock.java
package com.learning.leetcode.arrays.easy;

import java.util.Arrays;

public class BestTimeBuySellStock {
    public static void main(String[] args) {
        int[] prices = {7, 1, 5, 3, 6, 4};
        System.out.println("Max profit for " + Arrays.toString(prices) + " is " + getProfit(prices));
        prices = new int[]{7, 6, 4, 3, 1};
        System.out.println("Max profit for " + Arrays.toString(prices) + " is " + getProfit(prices));
    }

    private static int getProfit(int[] prices){
        int minPrice = Integer.MAX_VALUE;
        int profit = 0;
        for (int price : prices) {
            minPrice = Math.min(price, minPrice);
            profit = Math.max(profit, price - minPrice);
        }
        return profit;
    }
}
